### PR TITLE
Response generator responses with multimedia elements

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- added yaml support for responses. example of a `responses.yml` with the format as it is currently implemented:
```yaml
responses:
  chitchat/ask_weather:
  - text: Where do you want to check the weather?
    buttons:
    - title: Current location
      payload: here
    - title: Other place
      payload: other_location

  chitchat/ask_name:
  - text: my name is Sara, Rasa's documentation bot!
    image: "https://i.imgur.com/nGF1K8f.jpg"

```
- responses model only gets trained on the `text` part of the first `response`.
- allows users to use multimedia elements (images, buttons, ...) in response templates
- the format (and the parser) is the same as we use for utter templates in the domain
- this includes the changes to port the moodbot demo to the new file format (needed an example...)

**Status (please check what you already did)**:
- [x] handle case where the `response.text` is empty -> use the intent name instead
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tmbo** [comment]
@dakshvar22 are there any concerns regarding the data format?

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 1]
```diff
@@ -233,7 +233,7 @@ def is_markdown_story_file(file_path: Text) -> bool:
         """
         suffix = PurePath(file_path).suffix
 
-        if suffix and suffix != MARKDOWN_FILE_EXTENSION:
+        if suffix not in MARKDOWN_FILE_EXTENSIONS:
```
Comment (tmbo) [inline rasa/core/training/story_reader/markdown_story_reader.py:236]: I think we should discuss if we keep the original behaviour: I changed it to keep it uniform across readers. The difference between the versions is that the original one would happily read files without an extension as story files, which seems arbitrary. Nevertheless, it is breaking backwards compatibility in an odd way, so kind of betting that no one relies on this odd behaviour 🤔

File: rasa/nlu/classifiers/diet_classifier.py [hunk 2]
```diff
@@ -1098,7 +1104,7 @@ def _load_model(
             data_signature=model_data_example.get_signature(),
             label_data=label_data,
             entity_tag_specs=entity_tag_specs,
-            config=meta,
+            config=copy.deepcopy(meta),
```
Comment (tmbo) [inline rasa/nlu/classifiers/diet_classifier.py:1107]: this fix was needed as this config gets passed to tensorflow which in place convert all values of the dictionary to tensorflow types (e.g. it will replace an ordinary python string with a tensorflow specific string). to avoid tf modifying this original dictionary we need to pass in a copy...

**Comment by dakshvar22** [comment]
@tmbo  Data format looks good. Just two comments(maybe you have thought of these already):
1. In contrast to usual response templates, there cannot be multiple `text` attributes for one retrieval intent. For example, this would be invalid:
```
responses:
  chitchat/ask_weather:
  - text: Where do you want to check the weather?
    buttons:
    - title: Current location
      payload: here
    - title: Other place
      payload: other_location
 - text: Do you want to know the weather for today?
    buttons:
    - title: Today
      payload: here
    - title: Whole week
      payload: other_location
```
This is primarily because the response selector can only be trained with one ground truth label. Supporting multiple possible responses should be left out of scope for now IMO.

2. I am not sure of this but the `text` attribute of any response template is optional, right? For example, is this valid? - 
```
responses:
  chitchat/ask_weather:
  - buttons:
    - title: Current location
      payload: here
    - title: Other place
      payload: other_location
```
If it is valid, then we should use the full name of the retrieval intent(for e.g. - `chitchat/ask_weather`) as the proxy for the `text` attribute(just for model training)

**Comment by tmbo** [comment]
> In contrast to usual response templates, there cannot be multiple text attributes for one retrieval intent. For example, this would be invalid:

just because the ML model can't be trained with multiple of these, it doesn't mean this needs to be invalid. e.g. we could do the same as we do for utterance templates and just select a random one.

**Comment by tmbo** [comment]
> I am not sure of this but the text attribute of any response template is optional, right? For example, is this valid? -

Yes I think that is valid. I like the solution, let's use the intent name in that case

**Comment by dakshvar22** [comment]
> just because the ML model can't be trained with multiple of these, it doesn't mean this needs to be invalid. e.g. we could do the same as we do for utterance templates and just select a random one.
Yes, we can do that but which one would the model be trained on? It has to be just one and taking a random one for training, doesn't seem like a good option. If we always take the first response as the training response would it introduce any confusion/problems in the UX?(I am not very clear on this myself) For example, model would need to be retrained if the first response was swapped with the last.

**Comment by tmbo** [comment]
well, we can train the model on either the first, all of them or any of them: whatever you prefer.

I think from a UX perspective it is easier to explain that the format is exactly the same as the one for utterance templates. The model gets retrained on any change to the responses at the moment, I think from the training perspective it is pretty consistent.

File: rasa/nlu/training_data/training_data.py [hunk 3]
```diff
@@ -232,27 +236,21 @@ def sort_regex_features(self) -> None:
...
-                    )
+            if example.get(RESPONSE_KEY_ATTRIBUTE) is None:
+                continue
+
+            # look for corresponding bot utterance
+            story_lookup_intent = example.get_combined_intent_response_key()
+            assistant_utterances = self.responses.get(story_lookup_intent, [])
+            if assistant_utterances:
+                # selecting only first assistant utterance for now
+                example.set(RESPONSE, assistant_utterances[0].get("text"))
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:253]: Can we re-use an existing constant like `nlu.constants.TEXT` or define a new one?

File: rasa/nlu/training_data/training_data.py [hunk 4]
```diff
@@ -285,13 +295,21 @@ def nlu_as_yaml(self) -> Text:
...
     def persist_nlu(self, filename: Text = DEFAULT_TRAINING_DATA_OUTPUT_PATH):
 
-        if filename.endswith("json"):
+        if Path(filename).suffix in JSON_FILE_EXTENSIONS:
             rasa.nlu.utils.write_to_file(filename, self.nlu_as_json(indent=2))
-        elif filename.endswith("md"):
+        elif Path(filename).suffix in MARKDOWN_FILE_EXTENSIONS:
+            rasa.nlu.utils.write_to_file(filename, self.nlu_as_markdown())
+        elif Path(filename).suffix in YAML_FILE_EXTENSIONS:
             rasa.nlu.utils.write_to_file(filename, self.nlu_as_markdown())
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:313]: I'd also add a test which checks the correct persisting behavior 
```suggestion
            rasa.nlu.utils.write_to_file(filename, self.nlu_as_yaml())
```

File: rasa/nlu/training_data/training_data.py [hunk 5]
```diff
@@ -285,13 +295,21 @@ def nlu_as_yaml(self) -> Text:
...
-        return RasaYAMLWriter().dumps(self)
+        # avoid dumping NLG data (responses). this is a workaround until we
+        # can remove the distinction between nlu & nlg when converting to a string
+        # (so until after we remove markdown support)
+        no_responses_training_data = copy.copy(self)
+        no_responses_training_data.responses = {}
+
+        return RasaYAMLWriter().dumps(no_responses_training_data)
 
     def persist_nlu(self, filename: Text = DEFAULT_TRAINING_DATA_OUTPUT_PATH):
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:306]:
```suggestion
    def persist_nlu(self, filename: Text = DEFAULT_TRAINING_DATA_OUTPUT_PATH) -> None:
```

File: rasa/nlu/training_data/training_data.py (hunk 3 above)
Comment (dakshvar22) [inline rasa/nlu/training_data/training_data.py:253]: @tmbo I haven't read through the changes in the PR, but i am assuming this line means that we always take the text of the first response for model training in case there are multiple?

File: rasa/nlu/training_data/training_data.py [hunk 6]
```diff
@@ -384,6 +415,16 @@ def validate(self) -> None:
                     f"this the training may fail."
                 )
 
+        # emit warnings for response intents without a response template
+        for example in self.training_examples:
+            if example.get(RESPONSE_KEY_ATTRIBUTE) and not example.get(RESPONSE):
+                raise_warning(
+                    f"Your training data contains an example '{example.text[:20]}...' "
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:422]: optional: use a constant for `20`

File: rasa/nlu/training_data/training_data.py [hunk 7]
```diff
@@ -394,44 +435,38 @@ def train_test_split(
...
-    def build_nlg_stories_from_examples(examples) -> Dict[Text, list]:
+    def _needed_responses_for_examples(
+        self, examples: List[Message]
+    ) -> Dict[Text, List[Dict[Text, Any]]]:
+        """Get all responses used in any of the examples."""
 
-        nlg_stories = {}
+        responses = {}
         for ex in examples:
             if ex.get(RESPONSE_KEY_ATTRIBUTE) and ex.get(RESPONSE):
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:466]: can no longer happen due to the new validation, can it?
```suggestion
            if ex.get(RESPONSE_KEY_ATTRIBUTE):
```

File: rasa/nlu/training_data/training_data.py [hunk 8]
```diff
@@ -394,44 +435,38 @@ def train_test_split(
...
-        train_nlg_stories = self.build_nlg_stories_from_examples(train)
-        test_nlg_stories = self.build_nlg_stories_from_examples(test)
-        return test_nlg_stories, train_nlg_stories
-
-    @staticmethod
-    def build_nlg_stories_from_examples(examples) -> Dict[Text, list]:
+    def _needed_responses_for_examples(
+        self, examples: List[Message]
+    ) -> Dict[Text, List[Dict[Text, Any]]]:
+        """Get all responses used in any of the examples."""
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:462]: optional: Let's either add a full docstring according to our Python conventions or None

File: tests/nlu/training_data/test_training_data.py [hunk 9]
```diff
@@ -122,12 +122,12 @@ def test_demo_data(files):
     td = training_data_from_paths(files, language="en")
```
Comment (wochinge) [inline tests/nlu/training_data/test_training_data.py:122]: do we have a test for the new validation of `responses` in `TrainingData`?

File: tests/nlu/selectors/test_selectors.py [hunk 10]
```diff
@@ -38,7 +38,6 @@ def test_train_selector(pipeline, component_builder, tmpdir):
     td = load_data("data/examples/rasa/demo-rasa.md")
     td_responses = load_data("data/examples/rasa/demo-rasa-responses.md")
     td = td.merge(td_responses)
-    td.fill_response_phrases()
```
Comment (wochinge) [inline tests/nlu/selectors/test_selectors.py:41]: awesome moving this to the constructor 🚀

File: rasa/nlu/training_data/training_data.py (hunk 3 above)
Comment (tmbo) [reply rasa/nlu/training_data/training_data.py:253]: yes, the machine learning part is the same as it was before

Comment (dakshvar22) [reply rasa/nlu/training_data/training_data.py:253]: Okay, do you think we should still run the model regression tests before merging it?
The benchmarking datasets aren't in yaml format, nor do they contain any multimedia elements though.

File: rasa/nlu/training_data/training_data.py (hunk 6 above)
Comment (tmbo) [reply rasa/nlu/training_data/training_data.py:422]: I think it's fine to keep the 20: it is readable and just for logging. since it is very obvious what is happening, naming this random constant will not increase readability.

File: tests/nlu/training_data/test_training_data.py (hunk 9 above)
Comment (tmbo) [reply tests/nlu/training_data/test_training_data.py:122]: isn't that what https://github.com/RasaHQ/rasa/pull/6323/files#diff-91cfb497152a74c0719d07aaae3db53dR304 does? there is nothing more to test apart from the key being present and not empty, the validation isn't more specific.

File: rasa/nlu/training_data/training_data.py (hunk 3 above)
Comment (tmbo) [reply rasa/nlu/training_data/training_data.py]: don't think that is necessary since I didn't touch the ML parts, the input data is still exactly the same (tested in unit tests).

File: rasa/server.py [hunk 11]
```diff
@@ -835,6 +855,13 @@ async def evaluate_intents(request: Request) -> HTTPResponse:
             "evaluate your model.",
         )
 
+        if request.headers.get("Content-type") == YAML_CONTENT_TYPE:
+            test_data = _training_payload_from_yaml(request).get("training_files")
+        else:
+            test_data = rasa.utils.io.create_temporary_file(
+                request.body, mode="w+b", suffix=".md"
+            )
+
```
Comment (wochinge) [inline rasa/server.py:864]: optional: extract to function and de-duplicate as it's the same as above

File: rasa/server.py [hunk 12]
```diff
@@ -811,14 +825,20 @@ async def evaluate_stories(request: Request) -> HTTPResponse:
             "evaluate your model.",
         )
 
-        stories = rasa.utils.io.create_temporary_file(request.body, mode="w+b")
+        if request.headers.get("Content-type") == YAML_CONTENT_TYPE:
+            test_data = _training_payload_from_yaml(request).get("training_files")
```
Comment (wochinge) [inline rasa/server.py:829]: 🤠

File: rasa/server.py [hunk 13]
```diff
@@ -1177,14 +1203,12 @@ def _model_output_directory(save_to_default_model_directory: bool) -> Text:
 
 
 def _validate_yaml_training_payload(yaml_text: Text) -> None:
-    # TODO: Use NLU schema validation once it's merged
     try:
-        rasa.utils.io.read_yaml(yaml_text)
-    except Exception:
+        RasaYAMLReader.validate(yaml_text)
```
Comment (wochinge) [inline rasa/server.py:1207]: 🙏

File: tests/test_server.py [hunk 14]
```diff
@@ -399,16 +401,23 @@ def test_train_nlu_success(
     default_nlu_data: Text,
     default_domain_path: Text,
 ):
-    with ExitStack() as stack:
-        domain_file = stack.enter_context(open(default_domain_path))
-        config_file = stack.enter_context(open(default_stack_config))
-        nlu_file = stack.enter_context(open(default_nlu_data))
+    domain_data = rasa_utils.io.read_yaml_file(default_domain_path)
```
Comment (wochinge) [inline tests/test_server.py:404]: How about using `parametrize` or adding another test to make sure it still works with markdown?

File: rasa/nlu/training_data/training_data.py [hunk 15]
```diff
@@ -272,6 +271,18 @@ def nlg_as_markdown(self) -> Text:
 
         return NLGMarkdownWriter().dumps(self)
 
+    def nlg_as_yaml(self) -> Text:
+        """Generates yaml representation of the response phrases (NLG) of
+        TrainingData."""
```
Comment (wochinge) [inline rasa/nlu/training_data/training_data.py:276]: let's also describe the return type in the docstring

File: rasa/nlu/selectors/response_selector.py [hunk 16]
```diff
@@ -323,6 +326,14 @@ def preprocess_train_data(self, training_data: TrainingData) -> RasaModelData:
 
         return model_data
 
+    def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
+        """Given a label return the full response based on the labels id."""
```
Comment (wochinge) [inline rasa/nlu/selectors/response_selector.py:330]: let's either add a full docstring or none

File: rasa/nlu/selectors/response_selector.py [hunk 17]
```diff
@@ -323,6 +326,14 @@ def preprocess_train_data(self, training_data: TrainingData) -> RasaModelData:
...
+        for key, responses in self.responses.items():
+            for response in responses:
+                if hash(response.get("text", "")) == label.get("id"):
```
Comment (wochinge) [inline rasa/nlu/selectors/response_selector.py:333]: let's use the existing constant

File: rasa/nlu/classifiers/diet_classifier.py [hunk 18]
```diff
@@ -880,6 +881,7 @@ def _predict_label(
         # if X contains all zeros do not predict some label
         if label_ids.size > 0:
             label = {
+                "id": hash(self.index_label_id_mapping[label_ids[0]]),
```
Comment (wochinge) [inline rasa/nlu/classifiers/diet_classifier.py:884]: for my understanding: That's the hash of the intent name then, correct?

File: rasa/nlu/selectors/response_selector.py (hunk 17 above)
Comment (wochinge) [reply rasa/nlu/selectors/response_selector.py:333]: uff. the `text` of a `response` is the `RESPONSE` of a `Message`, correct? Maybe makes sense to add a comment

File: tests/nlu/training_data/formats/test_rasa_yaml.py [hunk 19]
```diff
@@ -280,3 +282,97 @@ def test_minimal_yaml_nlu_file(tmp_path):
...
+
+    with pytest.raises(InvalidDomain):
+        reader.reads(responses_yml)
+
+
+def test_nlg_multimedia_load_dump_roundtrip():
+    responses_yml = textwrap.dedent(
+        """
+      responses:
+        chitchat/ask_weather:
```
Comment (wochinge) [inline tests/nlu/training_data/formats/test_rasa_yaml.py:360]: optional: add another response to make sure the order isn't changed when dumping

File: docs/docs/chitchat-faqs.mdx [hunk 20]
```diff
@@ -26,65 +26,70 @@ In this section we're going to cover the following topics:
...
+```yaml title="config.yml"
 policies:
 - name: MemoizationPolicy
   max_history: 1
 - name: MappingPolicy
 ```
 
-:::note
-The MappingPolicy is there because it handles the logic of the `/restart` intent,
+:::note MappingPolicy
```
Comment (wochinge) [inline docs/docs/chitchat-faqs.mdx:62]: should probably be part of different PR: they should use the RulePolicy instead of the MappingPolicy

File: rasa/nlu/schemas/nlu.yml [hunk 21]
```diff
@@ -46,6 +46,10 @@ mapping:
         lookup:
           type: "str"
         examples: *examples_anchor
+  responses:
+    type: "map"
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/nlu/schemas/nlu.yml:50]: I don't think that is possible since the keys are the intent names, e.g. `chitchat/ask_name` and they would need to be known to further validate.

I think in the future we should consider moving to strict yaml instead of normal yaml + validation

File: rasa/nlu/training_data/lookup_tables_parser.py [hunk 22]
```diff
@@ -1,8 +1,8 @@
-from typing import Text, List, Dict
+from typing import Any, Text, List, Dict
 
 
 def add_item_to_lookup_tables(
-    title: Text, item: Text, existing_lookup_tables: List[Dict[Text, List[Text]]]
+    title: Text, item: Text, existing_lookup_tables: List[Dict[Text, Any]]
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/nlu/training_data/lookup_tables_parser.py:5]: this failed in my pytype, but I can change it back

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 1 above)
Comment (wochinge) [reply rasa/core/training/story_reader/markdown_story_reader.py:239]: Should we add this to the changelog? `Training data form in markdown format` has to have the file suffix `.md` from now on`?

File: rasa/nlu/classifiers/diet_classifier.py (hunk 18 above)
Comment (tmbo) [reply rasa/nlu/classifiers/diet_classifier.py:884]: nope hash of the text...but this will change again once #6392 gets merged...

File: rasa/nlu/selectors/response_selector.py (hunk 17 above)
Comment (tmbo) [reply rasa/nlu/selectors/response_selector.py:333]: yes, that is whats trained on. but as added in the other comment, this will change in the pr from daksh

File: rasa/nlu/selectors/response_selector.py (hunk 16 above)
Comment (tmbo) [reply rasa/nlu/selectors/response_selector.py:330]: I'll do so, but I am not sure this is a usefull practice: I am all with you that it is necessary for public methods. For private methods, I don't see why it is better to have no comment on a method than to add just a comment without args (assuming a helpful comment).

File: rasa/nlu/schemas/nlu.yml (hunk 21 above)
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/nlu/schemas/nlu.yml:50]: I think it should work using https://pykwalify.readthedocs.io/en/unstable/validation-rules.html#matching-rule 🤔

File: rasa/nlu/training_data/lookup_tables_parser.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/nlu/training_data/lookup_tables_parser.py:5]: what did pytype say?

File: docs/docs/chitchat-faqs.mdx (hunk 20 above)
Comment (tmbo) [reply docs/docs/chitchat-faqs.mdx:62]: yes, but that is separate

File: rasa/nlu/classifiers/diet_classifier.py (hunk 18 above)
Comment (wochinge) [reply rasa/nlu/classifiers/diet_classifier.py:884]: That's within the `DietClassifier` so I think it's the intent, no? Anyway, not important if get's changed by Daksh's PR

File: rasa/nlu/selectors/response_selector.py (hunk 16 above)
Comment (wochinge) [reply rasa/nlu/selectors/response_selector.py:330]: I think that's something which we should discuss as part of the coding conventions. I'd argue that 
- describing parameters + return type isn't that much of an effort
- then the comment is even more helpful
- it's more consistent

File: rasa/nlu/schemas/nlu.yml (hunk 21 above)
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/nlu/schemas/nlu.yml:50]: ah good point :+1: will add that in a separate PR as it needs to be added for the domain as well (I just copied it from there)

**Comment by tmbo** [comment]
thanks a lot @wochinge fur the super quick and thorough review :rocket:

File: tests/test_server.py (hunk 14 above)
Comment (tmbo) [reply tests/test_server.py:404]: I took a closer look and there are already a couple of other methods in there that test the same endpoint with markdown files. since this actually trains a model and takes time, I'd rather avoid adding more training runs. what do you think?

Comment (wochinge) [reply tests/test_server.py:404]: Yes, but then I'd maybe rename the tests to make it clear that this testing the MD support. Otherwise somebody might come, adapt the test, and suddenly all MD testing is gone.

Another quick way would be to extract the payload extraction to a separate function and test that using `parametrize`.

File: rasa/nlu/selectors/response_selector.py (hunk 16 above)
Comment (tmbo) [reply rasa/nlu/selectors/response_selector.py]: mhm, right thinking about it I agree :+1:

**Comment by wochinge** [comment]
@tmbo During the review I noticed the close coupling of training data objects (`Domain`, `TrainingData`) with the different `Writer` classes.

What do think about having a pendant to `TrainingDataImporter` to 
- support other file format implementations (e.g. implementations of the community, adapters to other frameworks)
- clear separation of concerns / classes don't do multiple things at the same time
- easier testability
- DRYer (each class is doing some sort of `if endswith "json" elif "endswith md"` currently)

**Comment by tmbo** [comment]
yes @wochinge I think that is a good idea :+1:
